File: fragments/postgres_db/postgres_crud.py
from typing import Optional, Tuple, List, Literal
import psycopg
from fragments.env_vars import get_env_var

CryptoCoin = Tuple[int, str, str, int]


def get_connection_string() -> str:
    connection_string = "postgres://postgres:pgpass@localhost:5432"

    if get_env_var("CI") is None:
        return connection_string

    return connection_string.replace("localhost", "postgres")


def init_db() -> psycopg.Connection:
    connection = psycopg.connect(get_connection_string())
    cursor = connection.cursor()
    cursor.execute(
        """
    CREATE TABLE IF NOT EXISTS crypto_coins (
      id SERIAL PRIMARY KEY,
      ticker VARCHAR(8) NOT NULL UNIQUE,
      name VARCHAR(30) NOT NULL,
      launched SMALLINT NOT NULL
    );
    INSERT INTO crypto_coins (ticker, name, launched) VALUES
    ('BTC', 'Bitcoin', 2009),
    ('ETH', 'Ethereum', 2015),
    ('SOL', 'Solana', 2020)
    ON CONFLICT (ticker) DO NOTHING; 
    """
    )
    connection.commit()

    return connection


def get_item_by_ticker(ticker: str) -> Optional[CryptoCoin]:
    connection = init_db()
    cursor = connection.execute("SELECT * FROM crypto_coins WHERE ticker = %s LIMIT 1", (ticker,))
    connection.close()

    return cursor.fetchone()


def get_items_after_launch_year(launch_year: int) -> List[CryptoCoin]:
    connection = init_db()
    cursor = connection.execute("SELECT * FROM crypto_coins WHERE launched > %s", (launch_year,))
    connection.close()

    return cursor.fetchall()


def get_all_items() -> List[CryptoCoin]:
    connection = init_db()
    cursor = connection.execute("SELECT * FROM crypto_coins ORDER BY launched DESC")
    connection.close()

    return cursor.fetchall()


def add_item(coin: Tuple[str, str, int]) -> Literal["ok"]:
    ticker = coin[0]
    name = coin[1]
    launched = coin[2]
    connection = init_db()
    connection.execute(
        "INSERT INTO crypto_coins (ticker, name, launched) VALUES(%s, %s, %s) ON CONFLICT (ticker) DO NOTHING;",
        (
            ticker,
            name,
            launched,
        ),
    )
    connection.commit()
    connection.close()

    return "ok"


def remove_item(ticker: str) -> Optional[CryptoCoin]:
    connection = init_db()
    cursor = connection.execute(
        "DELETE FROM crypto_coins WHERE ticker = %s RETURNING ticker, name, launched",
        (ticker,),
    )
    connection.commit()
    connection.close()

    return cursor.fetchone()


def update_item(coin: Tuple[str, str, int]) -> Optional[CryptoCoin]:
    ticker = coin[0]
    name = coin[1]
    launched = coin[2]
    connection = init_db()
    cursor = connection.execute(
        "UPDATE crypto_coins SET name = %s, launched = %s WHERE ticker = %s RETURNING ticker, name, launched",
        (
            name,
            launched,
            ticker,
        ),
    )
    connection.commit()
    connection.close()

    return cursor.fetchone()

File: fragments/postgres_db/postgres_crud.go
package postgres_crud

import (
	"context"
	"ff/env_vars"
	"fmt"
	"log"
	"strings"

	"github.com/jackc/pgx/v5"
)

type CryptoCoin struct {
	id       int
	ticker   string
	name     string
	launched int
}

type CryptoCoinWithoutId struct {
	Ticker   string
	Name     string
	Launched int
}

var ctx = context.Background()

func getConnectionString() string {
	localhost := "postgres://postgres:pgpass@localhost:5432"
	host := env_vars.GetEnvVar("POSTGRES_HOST")

	if len(host) == 0 {
		return localhost
	} else {
		return strings.Replace(localhost, "localhost", host, 1)
	}
}

func initDb() *pgx.Conn {
	conn, err := pgx.Connect(ctx, getConnectionString())
	if err != nil {
		log.Fatal("Unable to connect to Postgres", err)
	}
	_, err = conn.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS crypto_coins (
      id SERIAL PRIMARY KEY,
      ticker VARCHAR(8) NOT NULL UNIQUE,
      name VARCHAR(30) NOT NULL,
      launched SMALLINT NOT NULL
    );
    INSERT INTO crypto_coins (ticker, name, launched) VALUES
    ('BTC', 'Bitcoin', 2009),
    ('ETH', 'Ethereum', 2015),
    ('SOL', 'Solana', 2020)
    ON CONFLICT (ticker) DO NOTHING;
	`)
	if err != nil {
		log.Fatal("Error executing setup queries", err)
	}

	return conn
}

func GetItemByTicker(ticker string) (*CryptoCoin, error) {
	var coin CryptoCoin
	conn := initDb()
	defer conn.Close(ctx)
	err := conn.QueryRow(ctx,
		"SELECT * FROM crypto_coins WHERE ticker = $1 LIMIT 1",
		ticker,
	).Scan(&coin.id, &coin.ticker, &coin.name, &coin.launched)

	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("query failed: %w", err)
	}

	return &coin, nil
}

func GetItemsAfterLaunchYear(launchYear int) ([]CryptoCoin, error) {
	var coins []CryptoCoin
	conn := initDb()
	defer conn.Close(ctx)
	rows, queryErr := conn.Query(ctx, "SELECT * FROM crypto_coins WHERE launched > $1", launchYear)
	if queryErr != nil {
		return nil, fmt.Errorf("get items after launch year query failed: %w", queryErr)
	}

	defer rows.Close()

	for rows.Next() {
		var coin CryptoCoin
		err := rows.Scan(&coin.id, &coin.ticker, &coin.name, &coin.launched)
		if err != nil {
			return nil, fmt.Errorf("Error scanning row: %w", err)
		}
		coins = append(coins, coin)
	}

	if rows.Err() != nil {
		return nil, fmt.Errorf("row iteration error: %w", rows.Err())
	}

	return coins, nil
}

func GetAllItems() ([]CryptoCoin, error) {
	var coins []CryptoCoin
	conn := initDb()
	defer conn.Close(ctx)
	rows, queryErr := conn.Query(ctx, "SELECT * FROM crypto_coins ORDER BY launched DESC")
	if queryErr != nil {
		return nil, fmt.Errorf("get all items failed: %w", queryErr)
	}

	defer rows.Close()

	for rows.Next() {
		var coin CryptoCoin
		err := rows.Scan(&coin.id, &coin.ticker, &coin.name, &coin.launched)
		if err != nil {
			return nil, fmt.Errorf("error scanning row: %w", err)
		}
		coins = append(coins, coin)
	}

	if rows.Err() != nil {
		return nil, fmt.Errorf("row iteration error: %w", rows.Err())
	}

	return coins, nil
}

func CreateItem(coin CryptoCoinWithoutId) (string, error) {
	var query = "INSERT INTO crypto_coins (ticker, name, launched) VALUES($1, $2, $3) ON CONFLICT (ticker) DO NOTHING"
	conn := initDb()
	defer conn.Close(ctx)
	_, err := conn.Exec(ctx, query, coin.Ticker, coin.Name, coin.Launched)
	if err != nil {
		return "", fmt.Errorf("create item failed: %w", err)
	}

	return "ok", nil
}

func UpdateItem(coin CryptoCoinWithoutId) (*CryptoCoin, error) {
	var result CryptoCoin
	var query = "UPDATE crypto_coins SET name = $1, launched = $2 WHERE ticker = $3 RETURNING *"
	conn := initDb()
	defer conn.Close(ctx)
	err := conn.QueryRow(ctx,
		query,
		coin.Name, coin.Launched, coin.Ticker,
	).Scan(&result.id, &result.ticker, &result.name, &result.launched)

	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, err
		}
		return nil, fmt.Errorf("update item query failed: %w", err)
	}

	return &result, nil
}

func DeleteItem(ticker string) (*CryptoCoin, error) {
	var coin CryptoCoin
	var query = "DELETE FROM crypto_coins WHERE ticker = $1 RETURNING *"
	conn := initDb()
	defer conn.Close(ctx)
	err := conn.QueryRow(ctx,
		query,
		ticker,
	).Scan(&coin.id, &coin.ticker, &coin.name, &coin.launched)

	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("delete item query failed: %w", err)
	}

	return &coin, nil
}
